### PR TITLE
Add prototype of v2 controller framework

### PR DESCRIPTION
**Pull request by kuujo**

### CODE REVIEW

File: pkg/controller/v2/directive.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	go controller.enqueue(r.request)
+}
+
+type Fail[I ID] struct {
+	request Request[I]
+	Error   error
+}
+
+func (f *Fail[I]) Do(controller *Controller[I]) {
+	controller.Log.Warnf("Reconciliation of %s failed: %s", f.request.ID, f.Error.Error())
```
Comment (adibrastegarnia) [inline pkg/controller/v2/directive.go:50]: should we change these logs to use Warnw and make it structured?

Comment (kuujo) [reply pkg/controller/v2/directive.go]: done

File: pkg/controller/v2/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+// RetryAfter retries a reconciliation request after a delay
+type RetryAfter[I ID] struct {
+	*Retry[I]
+	delay time.Duration
+}
+
+// Do executes the controller directive
+func (r *RetryAfter[I]) Do(controller *Controller[I]) {
+	controller.Log.Debugw("Reconciliation of %s failed: %s. Retrying after %s", "Request.ID", r.request.ID, "Error", r.Error.Error(), "Delay", r.delay)
```
Comment (adibrastegarnia) [inline pkg/controller/v2/reconciler.go:159]: You still included formatting options here. perhaps fix this log.

File: pkg/controller/v2/reconciler.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+// RetryAt retries a reconciliation request at a specific time
+type RetryAt[I ID] struct {
+	*Retry[I]
+	t time.Time
+}
+
+// Do executes the controller directive
+func (r *RetryAt[I]) Do(controller *Controller[I]) {
+	controller.Log.Debugw("Reconciliation of %s failed: %s. Retrying at %s", "Request.ID", r.request.ID, "Error", r.Error.Error(), "Time", r.t)
```
Comment (adibrastegarnia) [inline pkg/controller/v2/reconciler.go:173]: same here

**Review by adibrastegarnia — APPROVED** [review]

**Comment by adibrastegarnia** [comment]
bump the version that we can use it

**Comment by kuujo** [comment]
done

**Pull request merged**